process/ingest.go

```diff
@@ -0,0 +1,143 @@
+package process
+
+import (
+	"errors"
+	"github.com/wptide/pkg/message"
+	"github.com/wptide/pkg/source"
+	"github.com/wptide/pkg/source/zip"
+	"crypto/sha256"
+	"encoding/base64"
+	"github.com/wptide/pkg/log"
+)
+
+// Ingest defines the structure for our Ingest process.
+type Ingest struct {
+	Process                           // Inherits methods from Process.
+	In         <-chan message.Message // Expects a message channel as input.
+	Out        chan Processor         // Send results to an output channel.
+	TempFolder string                 // Path to a temp folder where files will be extracted.
+	srcMgr     source.Source          // Responsible for getting the code to audit.
+}
+
+func (ig *Ingest) Run() (<-chan error, error) {
+	// If we don't have a temp folder, then we need a fatal.
+	if ig.TempFolder == "" {
+		return nil, errors.New("no temp folder provided for processes")
+	}
+	if ig.In == nil {
+		return nil, errors.New("no message channel to ingest")
+	}
+	if ig.Out == nil {
+		return nil, errors.New("requires a next process")
+	}
+
+	errc := make(chan error, 1)
+
+	go func() {
+		defer close(errc)
+		for {
+			select {
+			case msg := <-ig.In:
+
+				// If message is invalid, skip it, but keep listening on the channel.
+				if err := validateMessage(msg); err != nil {
+					// Pass the error up the error channel.
+					errc <- err
+					// break so that the message doesn't get passed along.
+					break
+				}
+
+				// Get the original message.
+				ig.SetMessage(msg)
+
+				// Run the process.
+				// If processing produces an error send it up the error channel.
+				if err := ig.process(); err != nil {
+					// Pass the error up the error channel.
+					errc <- err
+					// break so that the message doesn't get passed along.
+					break
+				}
+
+				// Send process to the out channel.
+				ig.Out <- ig
+			case <-ig.context.Done():
+				return
+			}
+		}
+
+	}()
+
+	return errc, nil
+}
+
+func (ig *Ingest) process() error {
+
+	ig.Result = make(map[string]interface{})
+
+	log.Log(ig.Message.Title, "Ingesting...")
+
+	// Set the source manager based on message.
+	switch source.GetKind(ig.Message.SourceURL) {
+	case "zip":
```

Comment:
Does this mean we currently only support ZIP as a source?

Reply:
Correct, we have only supported zip so far. I have not added any additional sources yet.

Reply:
The previous AS supports Git, and is trivial to add Subversion support but we hadn't yet.

Reply:
Fair call... In my head "previous" meant the `audit` package. Symfony is so far away from memory ;)   I need to finish the services first and will see if I have time to add additional source before I head off for the next couple of weeks.